### PR TITLE
MODINVOICE-69 Restrict creation of invoice, invoiceLine records based upon acquisitions unit

### PR DESCRIPTION
**Pull request by javalkar-varun**
## Purpose

https://issues.folio.org/browse/MODINVOICE-69

## Approach

- Restrict Creation of invoice and invoiceLines based upon the acquisitions unit memberships of the user and the acquisitions unit being assigned to the record.
- Used same approach described in [wiki](https://wiki.folio.org/display/FOLIJET/Acquisition+Units)
- Add unit tests
- Fixed code smells

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [x] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -18,9 +18,12 @@
           "methods": ["POST"],
           "pathPattern": "/invoice/invoices",
           "permissionsRequired": ["invoice.invoices.item.post"],
+          "permissionsDesired": ["invoice.acquisitions-units-assignments.item.post"],
```
Comment (piotr-kalashuk) [inline descriptors/ModuleDescriptor-template.json]: I think better to use common naming patter for such permissions like orders app has (`orders.acquisitions-units-assignments.assign`)

File: README.md [hunk 2]
```diff
@@ -11,6 +11,9 @@ This is the Invoice business logic module.
 
 ## Additional information
 
+NOTE: Only in case an acquisition unit has to be assigned to the Invoice, it is required that user should have an
+extra permission `invoice.acquisitions-units-assignments.item.post` to create an invoice.
```
Comment (KVupp) [inline README.md]: This permission is different from what is defined in AcqDesiredPermissions

File: src/main/java/org/folio/rest/impl/InvoicesImpl.java [hunk 3]
```diff
@@ -81,13 +87,19 @@ public void putInvoiceInvoicesById(String id, String lang, Invoice invoice, Map<
...
 
     // Validate incoming invoice first to avoid extra calls to other services if content is invalid
-    if (!invoiceHelper.validateIncomingInvoice(invoice)) {
-      asyncResultHandler.handle(succeededFuture(invoiceHelper.buildErrorResponse(422)));
-      return;
-    }
-
-    invoiceHelper.updateInvoice(invoice)
-      .thenAccept(ok -> asyncResultHandler.handle(succeededFuture(invoiceHelper.buildNoContentResponse())))
+    invoiceHelper.validateIncomingInvoice(invoice)
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/InvoicesImpl.java]: why is this change needed?

File: src/main/java/org/folio/rest/impl/ProtectionHelper.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      return CompletableFuture.completedFuture(null);
+    }
+  }
+
+  /**
+   * Check whether the user is a member of at least one group from which the related invoice belongs.
+   *
+   * @return list of unit ids associated with user.
+   */
+  private CompletableFuture<Void> verifyUserIsMemberOfInvoiceUnits(List<String> unitIdsAssignedToInvoice) {
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/ProtectionHelper.java]: The method name is a little confusing, can you please add acquistionsunits, so its evident

File: src/main/java/org/folio/rest/impl/InvoicesImpl.java (hunk 3 above)
Comment (javalkar-varun) [reply src/main/java/org/folio/rest/impl/InvoicesImpl.java]: Once `validateIncomingInvoice` return type changed to `CompletableFuture<Boolean>`, it starting erroring out with `The operator ! is undefined for the argument type(s) CompletableFuture<Boolean>`. So to fix it, I had to refactor to use CompletableFuture.

File: src/main/java/org/folio/invoices/rest/exceptions/HttpException.java [hunk 5]
```diff
@@ -17,8 +17,9 @@ public HttpException(int code, String message) {
   }
 
   public HttpException(int code, ErrorCodes errorCode) {
+    super(errorCode.getDescription());
     this.code = code;
-    this.error = errorCode.toError();
+    this.error = new Error().withCode(errorCode.getCode()).withMessage(errorCode.getDescription());
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/invoices/rest/exceptions/HttpException.java]: It looks to me that `errorCode.toError()` does absolutely the same stuff...

File: src/main/java/org/folio/invoices/utils/HelperUtils.java [hunk 6]
```diff
@@ -164,7 +164,9 @@ public static String getEndpointWithQuery(String query, Logger logger) {
     return future;
   }
 
-
+  public static String buildQuery(String query, Logger logger) {
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/invoices/utils/HelperUtils.java]: Such method is already available but named a bit differently: `getEndpointWithQuery`

Comment (javalkar-varun) [reply src/main/java/org/folio/invoices/utils/HelperUtils.java]: Sure, I'm working on resolving merge conflicts and will be refactoring most of the flies.
